native/thrift_struct_read_module.c

```diff
@@ -0,0 +1,1851 @@
+
+
+struct structs 
+popStructs()
+{
+    PyObject *structsStack;
+    PyObject *py_size;
+    Py_ssize_t structsTop;
+    PyObject *globalVars;
+    struct structs temp;
+        
+    globalVars = PyThreadState_GetDict();
+    
+    structsStack = PyDict_GetItemString(globalVars, "structsStack");
+    py_size = PyDict_GetItemString(globalVars, "structsStackTop");
+    if (!PyArg_Parse(py_size, "n", &structsTop)) {
+            return temp;
+    }
+
+    if (structsTop != -1) {
+        PyArg_ParseTuple(PyList_GetItem(structsStack, structsTop), "ii", &(temp.state), &(temp.last_fid));
+        structsTop--;
+        PyDict_SetItemString(globalVars, "structsStackTop", Py_BuildValue("n", structsTop));
+        return temp;
+    }
+    return temp;
+}
+
+int 
+popContainers()
+{
+    PyObject *ContainersStack;
+    PyObject *py_size;
+    Py_ssize_t containersTop;
+    PyObject *globalVars;
+    int temp;
+        
+    globalVars = PyThreadState_GetDict();
+    
+    ContainersStack = PyDict_GetItemString(globalVars, "containersStack");
+    py_size = PyDict_GetItemString(globalVars, "containersStackTop");
+    if (!PyArg_Parse(py_size, "n", &containersTop)) {
+            return temp;
+    }
+
+    if (containersTop != -1) {
+        PyArg_ParseTuple(PyList_GetItem(ContainersStack, containersTop), "i", &temp);
+        containersTop--;
+        PyDict_SetItemString(globalVars, "containersStackTop", Py_BuildValue("n", containersTop));
+        return temp;
+    }
+    return temp;
+}
+
+
+PyObject *thrift_struct_read(char*, char*, int, int, int, int, int);
+PyObject *read_field_value_C(char* ,char*, int, int, int, int, int, int);
+void readStructBeginC(char*, char*);
+void readStructEndC(char*, char*);
+int readFieldBeginC(char*, char*, char*, int*);
+void readFieldEndC(char*, char*);
+int readByteC(char*);
+int readBool(char*, char*);
+int readI16(char*, char*);
+int readI32(char*, char*);
+long int readI64(char*, char*);
+double readDouble(char*, char*);
+char *readString(char*, char*);
+char *readBinary(char*, char*);
+void readMapBegin(char*, char*, int*, int*, int*);
+void readMapEnd(char*, char*);
+void readSetBegin(char*, char*, int*, int*);
+void readSetEnd(char*,  char*);
+void readListBegin(char *, char*, int*, int*);
+void readListEnd(char*, char*);
+void skip(char*, char*, int);
+char *field_type_to_str(int);
+
+int readJson(char*);
+int peekJson(char*);
+void contextRead(char*);
+int escapeNum();
+void readJSONObjectStart(char*);
+void readJSONObjectEnd(char*);
+void readJSONArrayStart(char*);
+void readJSONArrayEnd(char*);
+void readJSONSyntaxChar(char*, char);
+int readJsonInteger(char*);
+void readJsonQuotes(char*);
+int readJsonNumericChars(char*);
+int JTypesToInt(char*);
+char *readJSONString(char*, int);
+double readJSONDouble(char*);
+void pushContext(int, int, int);
+void popContext();
+
+int readSize(char*);
+int readVarint(char*);
+int fromZigZag(int);
+int getTType(int);
+int readerZigZag(char*);
```

Comment:
all of these methods are private to this file, so they should be static. See: 

https://stackoverflow.com/questions/558122/what-is-a-static-function